crates/aiken-lang/src/tipo/environment.rs

```diff
@@ -1415,12 +1415,12 @@ impl<'a> Environment<'a> {
             && !(t1.is_function() || t2.is_function())
             && !(t1.is_generic() || t2.is_generic())
             && !(t1.is_string() || t2.is_string())
-            && !(t1.contains_opaque() || t2.contains_opaque())
+            && !t1.contains_opaque()
         {
             return Ok(());
         }
 
-        if allow_cast && (t1.contains_opaque() || t2.contains_opaque()) {
+        if allow_cast && t1.contains_opaque() {
```

Comment:
This only works fine until we reestablish that casting is only from Data and to Data not arbitrary types.

Reply:
ok this works nice